src/uct/ib/ud/base/ud_iface.c

```diff
@@ -813,14 +813,10 @@ uct_ud_send_skb_t *uct_ud_iface_ctl_skb_get(uct_ud_iface_t *iface)
 unsigned uct_ud_iface_dispatch_async_comps_do(uct_ud_iface_t *iface)
 {
     unsigned count = 0;
-    uct_ud_comp_desc_t *cdesc;
     uct_ud_send_skb_t *skb;
 
     ucs_queue_for_each_extract(skb, &iface->tx.async_comp_q, queue, 1) {
-        ucs_assert(!(skb->flags & UCT_UD_SEND_SKB_FLAG_RESENDING));
-        cdesc = uct_ud_comp_desc(skb);
-        uct_ud_iface_dispatch_comp(iface, cdesc->comp, cdesc->status);
-        uct_ud_skb_release(skb, 0);
+        uct_ud_iface_dispatch_ep_async_comp(iface, NULL, skb);
```

Comment:
maybe pass ep to uct_ud_iface_dispatch_async_comps_do? and remove uct_ud_ep_dispatch_async_comps

Reply:
the idea having `uct_ud_ep_dispatch_async_comps()` to not add overhead to `uct_ud_iface_dispatch_async_comps_do()`
having a check `NULL` for `EP` removes this branch, because the function is static inline

Reply:
IMO, it's not critical: most of the time, this queue would be empty. so the overhead of extra check is OK

Reply:
fixed